ping/ping6_common.c

```diff
@@ -104,8 +104,7 @@ unsigned int if_name2index(const char *ifname)
 int ping6_run(struct ping_rts *rts, int argc, char **argv, struct addrinfo *ai,
 	      struct socket_st *sock)
 {
-	int hold, packlen;
-	size_t i;
+	size_t hold, i, packlen;
```

Comment:
The change for hold from `int`' to '`size_t` seem to be suspicions. We do pass the hold to a `setsockopt()` later in the code in:

```
        hold = 1;
        if (setsockopt(sock->fd, IPPROTO_IPV6, IPV6_RECVERR, (const void *)&hold, sizeof hold))
                error(2, errno, "IPV6_RECVERR");
```

Does that still work as expected?

Also the ping6_common() misuses the hold for a return value, I guess that it would be cleaner to write it as:

```
diff --git a/ping/ping6_common.c b/ping/ping6_common.c
index eb3b0d2..8431d56 100644
--- a/ping/ping6_common.c
+++ b/ping/ping6_common.c
@@ -410,9 +410,9 @@ int ping6_run(struct ping_rts *rts, int argc, char **argv, struct addrinfo *ai,
 
        drop_capabilities();
 
-       hold = main_loop(rts, &ping6_func_set, sock, packet, packlen);
+       int ret = main_loop(rts, &ping6_func_set, sock, packet, packlen);
        free(packet);
-       return hold;
+       return ret;
 }
```

Reply:
> The change for `hold` from `int` to `size_t` seem to be suspicions. We do pass the hold to a `setsockopt()` later in the code in:

Very good catch. Strangely, ping did not warn even on non-zero exit code. I suppose `hold` must be changed as well, because it also stores value of `size_t`. But I created `int ret` and use it in separate commit.

Reply:
What about the part where we pass sizeof(hold) to the setsockopt() is that code okay with hold changed to size_t?

And looking at kernel sources it does:

```
                if (optlen >= sizeof(int)) {
                        if (copy_from_sockptr(&val, optval, sizeof(val)))
                                return -EFAULT;
                 }
```

So if we pass size bigger sizeof(int) kernel will take a first sizeof(int) bytes from the value and use that. Now on 64bit the sizeof(int) == 32 and sizeof(size_t) == 64. So we take only the half of the value, on little-endian things are fine, since we will take the lower half of the integer and the value will be fine. But on big-endian we will take the upper half of the integer and wrongly interpret the value as 0.

So I guess that we should add another patch that decouples the use of the hold from the setsockopt() calls as well.